llvm/lib/DebugInfo/LogicalView/Readers/LVIRReader.cpp

```diff
@@ -0,0 +1,2309 @@
+      constructTemplateTypeParameter(Function, TTP);
+    else if (const auto *TVP = dyn_cast<DITemplateValueParameter>(DN))
+      constructTemplateValueParameter(Function, TVP);
+  }
+
+  applySubprogramAttributes(Function, SP);
+
+  // Check if we are dealing with the Global Init/Cleanup Function.
+  if (SP->isArtificial() && SP->isLocalToUnit() && SP->isDefinition() &&
+      SP->getName().empty())
+    Function->setName(SP->getLinkageName());
+
+  return Function;
+}
+
+void LVIRReader::constructSubprogramArguments(LVScope *Function,
+                                              const DITypeRefArray Args) {
+  assert(Function && "Invalid logical element");
+  LLVM_DEBUG({
+    dbgs() << "\n[constructSubprogramArguments] DITypeRefArray\n";
+    for (unsigned i = 1, N = Args.size(); i < N; ++i) {
+      if (const DIType *Ty = Args[i])
+        Ty->dump();
+    }
+  });
+
+  for (unsigned I = 1, N = Args.size(); I < N; ++I) {
+    const DIType *Ty = Args[I];
+    LVElement *Parameter = nullptr;
+    if (Ty) {
+      // Create a formal parameter.
+      LVElement *ParameterType = getOrCreateType(Ty);
+      Parameter = createElement(dwarf::DW_TAG_formal_parameter);
+      if (Parameter) {
+        Parameter->setType(ParameterType);
+        if (Ty->isArtificial())
+          Parameter->setIsArtificial();
+      }
+    } else {
+      // Add an unspecified parameter.
+      Parameter = createElement(dwarf::DW_TAG_unspecified_parameters);
+    }
+    if (Parameter) {
+      Function->addElement(Parameter);
+      Parameter->setIsFinalized();
+    }
+  }
+}
+
+// DISubrange
+void LVIRReader::constructSubrange(LVScopeArray *Array, const DISubrange *GSR,
+                                   LVType *IndexType) {
+  assert(Array && "Invalid logical element");
+  assert(GSR && "Invalid metadata node.");
+  LLVM_DEBUG({
+    dbgs() << "\n[constructSubrange] DISubrange\n";
+    GSR->dump();
+  });
+
+  // The DISubrange can be shared between different arrays, when they are
+  // the same. We need to create independent logical elements for each one,
+  // as they are going to be added to different arrays.
+  if (LVTypeSubrange *Subrange =
+          static_cast<LVTypeSubrange *>(constructElement(GSR))) {
+    Subrange->setIsFinalized();
+    Array->addElement(Subrange);
+    Subrange->setType(IndexType);
+
+    int64_t Count = -1;
+    // If Subrange has a Count field, use it.
+    // Otherwise, if it has an upperboud, use (upperbound - lowerbound + 1),
+    // where lowerbound is from the LowerBound field of the Subrange,
+    // or the language default lowerbound if that field is unspecified.
+    if (auto *CI = dyn_cast_if_present<ConstantInt *>(GSR->getCount()))
+      Count = CI->getSExtValue();
+    else if (auto *UI =
+                 dyn_cast_if_present<ConstantInt *>(GSR->getUpperBound())) {
+      // Fortran uses 1 as the default lowerbound; other languages use 0.
+      int64_t Lowerbound = (moduleIsInFortran()) ? 1 : 0;
```

Comment:
It seems like being Fortran or not does not matter, except for the subrange lowerbound.  Should we store the default-array-lowerbound (instead of `LanguageIsFortran`)?